src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookSearchView.ts

```diff
@@ -0,0 +1,1332 @@
+	}
+
+	protected layoutBody(height: number, width: number): void {
+		super.layoutBody(height, width);
+		this.size = new dom.Dimension(width, height);
+		this.reLayout();
+	}
+
+	getControl() {
+		return this.tree;
+	}
+
+	isSlowSearch(): boolean {
+		return this.state === SearchUIState.SlowSearch;
+	}
+
+	hasSearchResults(): boolean {
+		return !this.viewModel.searchResult.isEmpty();
+	}
+
+	clearSearchResults(clearInput = true): void {
+		this.viewModel.searchResult.clear();
+		this.showEmptyStage(true);
+		if (this.contextService.getWorkbenchState() === WorkbenchState.EMPTY && !this.contextKeyService.getContextKeyValue('bookOpened')) {
+			this.showSearchWithoutFolderMessage();
+		}
+		this.viewModel.cancelSearch();
+		this.updateActions();
+
+		aria.status(nls.localize('ariaSearchResultsClearStatus', "The search results have been cleared"));
+	}
+
+	cancelSearch(focus: boolean = true): boolean {
+		if (this.viewModel.cancelSearch()) {
+			return true;
+		}
+		return false;
+	}
+
+	public selectTreeIfNotSelected(): void {
+		if (this.tree.getNode(null)) {
+			this.tree.domFocus();
+			const selection = this.tree.getSelection();
+			if (selection.length === 0) {
+				this.tree.focusNext();
+			}
+		}
+	}
+
+	searchInFolders(resources?: URI[]): void {
+		const folderPaths: string[] = [];
+		const workspace = this.contextService.getWorkspace();
+
+		if (resources) {
+			resources.forEach(resource => {
+				let folderPath: string | undefined;
+				if (this.contextService.getWorkbenchState() === WorkbenchState.FOLDER) {
+					// Show relative path from the root for single-root mode
+					folderPath = relativePath(workspace.folders[0].uri, resource); // always uses forward slashes
+					if (folderPath && folderPath !== '.') {
+						folderPath = './' + folderPath;
+					}
+				} else {
+					const owningFolder = this.contextService.getWorkspaceFolder(resource);
+					if (owningFolder) {
+						const owningRootName = owningFolder.name;
+
+						// If this root is the only one with its basename, use a relative ./ path. If there is another, use an absolute path
+						const isUniqueFolder = workspace.folders.filter(folder => folder.name === owningRootName).length === 1;
+						if (isUniqueFolder) {
+							const relPath = relativePath(owningFolder.uri, resource); // always uses forward slashes
+							if (relPath === '') {
+								folderPath = `./${owningFolder.name}`;
+							} else {
+								folderPath = `./${owningFolder.name}/${relPath}`;
+							}
+						} else {
+							folderPath = resource.fsPath; // TODO rob: handle on-file URIs
+						}
+					}
+				}
+
+				if (folderPath) {
+					folderPaths.push(folderPath);
+				}
+			});
+		}
+
+		if (!folderPaths.length || folderPaths.some(folderPath => folderPath === '.')) {
+			return;
+		}
+	}
+
+	public doSearch(query: ITextQuery, excludePatternText: string, includePatternText: string, triggeredOnType: boolean, searchWidget: NotebookSearchWidget): Thenable<void> {
```

Comment:
Use Promise<> not Thenable<>

Reply:
This method is over 144 lines long - can this not be broken up into manageable pieces?